woocommerce/woocommerce-gutenberg-products-block
# Google Analytics Integration - Action Hooks and initial tracking events

This is my first pass on the Google Analytics Integration within blocks. This requires the [Google Analytics Integration Extension](https://github.com/woocommerce/woocommerce-google-analytics-integration) to test functionality, setup with a G-X type ID for gtag support, and to test events are firing you can use the [Google Analytics Debugger.](https://chrome.google.com/webstore/detail/google-analytics-debugger/jnkmfdileelhofjcijamephohjechhna?hl=en) I am also logging events to the console to aid in development, but these will be removed before shipping.

The PHP side code includes the gtag script. There is a check to see if this script is already enqueued, but it will never be unless [this issue is resolved in the extension](https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/186). For now we work around this.

Events from blocks are triggered using the [`@wordpress/hooks`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-hooks/) package. Event names have the prefix `experimental__` (underscore prefix is not allowed).

A separate script with it's own entry point is included for the actual GA events. This hooks in using `addAction` and currently covers the following event:

- All Products Block View triggers a `view_item_list` event
- All Products Block add to cart button triggers an `add_to_cart` event
- Cart block remove item triggers a `remove_cart_item` event
- Cart block change qty triggers a custom `change_cart_quantity` event

I chose to trigger events at component level rather than hook level to offer more control over event names and how tracking is done without impacting other potential consumers of the hooks.

Closes #3934 
Part of #3935 

Seeking feedback before moving on to other events.

## How to test the changes in this Pull Request:

- Install the [Google Analytics Integration Extension](https://github.com/woocommerce/woocommerce-google-analytics-integration)
- Setup the extension with a G-X type ID from your GA account 
- Install [Google Analytics Debugger](https://chrome.google.com/webstore/detail/google-analytics-debugger/jnkmfdileelhofjcijamephohjechhna?hl=en) extension and enable it
- Go to the store and trigger some events using the All Products Block and Cart block. See list of events above.
- The console will show when events are fired, and the Google Analytics Debugger will also show logs
